File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string filter(string s, char open, char close)
    {
        int balance = 0;
        string ans = "";
        
        
        for(auto& c : s)
        {
            if(c == open) balance++;
            else if(c == close)
            {
                if(balance <= 0) continue;
                balance--;
            }
            ans += c;
        }
        
        return ans;
    }
    string minRemoveToMakeValid(string s) {
        string result = filter(s, '(', ')');
        reverse(result.begin(), result.end());
        result = filter(result, ')', '(');
        reverse(result.begin(), result.end());
        return result;
    }
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        
        
        def helper(s, open, close):
            balance = 0
            ans = ""

            for c in s:
                if c == open:
                    balance+=1
                elif c == close:
                    balance-=1
                if balance == -1:
                    balance = 0
                else:
                    ans += c
            return ans
        
        s = helper(s, "(", ")")
        s = s[::-1]
        s = helper(s, ")", "(")
        s = s[::-1]
        return s